File: Cubixcraft_v0.0.1.1-alpha_2020.04.19_src_VisualC++2019/Cubixcraft/font.cpp
#include "font.h"
#include "main.h"

GLuint fontTex = 0;
GLuint hFontLists = 0;

/*
 * BuildTexturedFont()
 * Загружает текстурный шрифт.
 * Текстура должна быть размером 256x256 и иметь формат PNG 32 bits с альфа-каналом.
 */
GLuint BuildTexturedFont(char* textureFileName) {
    fontTex = SOIL_load_OGL_texture(textureFileName,
        SOIL_LOAD_AUTO, SOIL_CREATE_NEW_ID, SOIL_FLAG_INVERT_Y);
    if (fontTex) {
        float w = 1.0f / 16.0f;
        hFontLists = glGenLists(256);
        glBindTexture(GL_TEXTURE_2D, fontTex);

        //создаём отдельные листы для каждой буквы
        for (int i = 0; i < 256; i++) {
            float cx = (i % 16) / 16.0f;
            float cy = (i / 16) / 16.0f;
            glNewList(hFontLists + i, GL_COMPILE);
            glBegin(GL_QUADS);
            glTexCoord2f(cx,     1.0f - cy - w);
            glVertex2i(0, 0);
            glTexCoord2f(cx + w, 1.0f - cy - w);
            glVertex2i(16, 0);
            glTexCoord2f(cx + w, 1.0f - cy);
            glVertex2i(16, 16);
            glTexCoord2f(cx,     1.0f - cy);
            glVertex2i(0, 16);
            glEnd();
            glTranslatef(symbolWidth, 0.0f, 0.0f);
            glEndList();
        }
        std::cout << "Loaded font: " << textureFileName << std::endl;
    }
    return fontTex;
}

/*
 * KillFont()
 * Выгружает текстуру шрифта и освобождает память.
 */
void KillFont() {
    if (hFontLists) {
        glDeleteLists(hFontLists, 256);
        hFontLists = 0;
    }
    if (fontTex) {
        glDeleteTextures(1, &fontTex);
        fontTex = 0;
    }
}

/*
 * glTextStart()
 * Устанавливает необходимые параметры для вывода текста и
 * переключает OpenGL в 2D-режим рисования.
 */
void glTextStart() {
    glEnable(GL_ALPHA_TEST);
    glAlphaFunc(GL_GREATER, 0.0f);
    glDisable(GL_DEPTH_TEST);

    glBindTexture(GL_TEXTURE_2D, fontTex);

    glMatrixMode(GL_PROJECTION);
    glPushMatrix();
    glLoadIdentity();

    glOrtho(0.0, (double)glScreenWidth, 0.0, (double)glScreenHeight, -1.0, 1.0);
    glMatrixMode(GL_MODELVIEW);
    glPushMatrix();
    glLoadIdentity();
}

/*
 * glTextEnd()
 * Завершить вывод текста и переключиться обратно в 3D-режим.
 */
void glTextEnd() {
    glMatrixMode(GL_PROJECTION);
    glPopMatrix();
    glMatrixMode(GL_MODELVIEW);
    glPopMatrix();

    glEnable(GL_DEPTH_TEST);
    glDisable(GL_ALPHA_TEST);
}

/*
 * glTextOut()
 * Выводит текст на экран.
 */
void glTextOut(int xPos, int yPos, const char* sText, int color) {
    glLoadIdentity();
    float r = (color >> 16 & 0xFF) / 255.0f;
    float g = (color >> 8  & 0xFF) / 255.0f;
    float b = (color       & 0xFF) / 255.0f;
    glColor3f(r, g, b);
    glTranslated(xPos, (double)glScreenHeight - 20.0 - yPos, 0.0);
    glListBase(hFontLists - 32);
    glEnable(GL_TEXTURE_2D);
    glCallLists(strlen(sText), GL_BYTE, sText);
    glDisable(GL_TEXTURE_2D);
}

/*
 * DrawTextBkg()
 * Рисует полупрозрачный прямоугольник.
 */
void DrawTextBkg(int xPos, int yPos, int xWidth, int color) {
    glEnable(GL_BLEND);
    glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
    float r = (color >> 16 & 0xFF) / 255.0f;
    float g = (color >>  8 & 0xFF) / 255.0f;
    float b = (color       & 0xFF) / 255.0f;
    glColor4f(r, g, b, 0.2f);
    glLoadIdentity();
    glBegin(GL_QUADS);
    glVertex3f((GLfloat)xPos,            (GLfloat)(glScreenHeight - yPos - 20.0f), 0.0f);
    glVertex3f((GLfloat)(xPos + xWidth), (GLfloat)(glScreenHeight - yPos - 20.0f), 0.0f);
    glVertex3f((GLfloat)(xPos + xWidth), (GLfloat)(glScreenHeight - yPos),         0.0f);
    glVertex3f((GLfloat)xPos,            (GLfloat)(glScreenHeight - yPos),         0.0f);
    glEnd();
    glDisable(GL_BLEND);
}

std::string FormatStr(char* sText, ...) {
    char text[256];
    va_list vl;
    va_start(vl, sText);
    vsprintf_s(text, sText, vl);
    va_end(vl);
    return text;
}

File: cubixcraft_v0.0.2-alpha_2020.05.16_java_src_IntelliJidea/src/game/cubixcraft/utils/Font.java
package game.cubixcraft.utils;

import game.cubixcraft.CubixCraft;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

import static org.lwjgl.opengl.GL11.*;

public class Font {

    private int imageWidth;
    private int imageHeight;
    private int textureId;
    private float fontSize = 16.0f;
    private float widthCoef = 1.5f;
    private float[] charWidths;

    public Font(String fileName) {
        /*
         * Загружает текстурный шрифт.
         * Текстура должна быть размером 256x256 и иметь формат PNG 32 bits с альфа-каналом.
         */
        System.out.println("Loading font...");
        URL url;
        if (fileName.startsWith("/")) {
           url = Font.class.getResource(fileName);
        }
        else
        {
            try {
                url = new File(fileName).toURI().toURL();
            } catch (MalformedURLException e) {
                e.printStackTrace();
                url = null;
            }
        }
        if (url == null) {
            System.out.println("Can not find " + fileName);
            throw new RuntimeException("Can not find " + fileName);
        }
        BufferedImage img;
        try {
            img = ImageIO.read(url);
        } catch (IOException e) {
            System.out.println("Can not load " + fileName);
            e.printStackTrace();
            throw new RuntimeException("Can not load " + fileName);
        }
        imageWidth = img.getWidth();
        imageHeight = img.getHeight();
        textureId = Textures.loadTexture(fileName);
        charWidths = new float[256];
    }

    public void start() {
        glEnable(GL_ALPHA_TEST);
        glAlphaFunc(GL_GREATER, 0.0f);
        glDisable(GL_DEPTH_TEST);

        glBindTexture(GL_TEXTURE_2D, textureId);

        glMatrixMode(GL_PROJECTION);
        glLoadIdentity();
        glOrtho(0.0, CubixCraft.windowWidth, CubixCraft.windowHeight, 0.0, -1.0, 1.0);
        glMatrixMode(GL_MODELVIEW);
        glLoadIdentity();
    }

   public void end() {
        glDisable(GL_TEXTURE_2D);
        glDisable(GL_ALPHA_TEST);
        glEnable(GL_DEPTH_TEST);
    }

    public void drawString(int xPos, int yPos, String textString, int color) {
        for (int i = 0; i < 256; i++) {
            charWidths[i] = fontSize / widthCoef;
        }
        charWidths['l'] = fontSize / 2.0f / widthCoef;
        float r = (color >> 16 & 0xFF) / 255.0f;
        float g = (color >>  8 & 0xFF) / 255.0f;
        float b = (color       & 0xFF) / 255.0f;
        glColor3f(r, g, b);
        char[] chars = textString.toCharArray();
        int glyphSize = imageWidth / 16;
        float xOffset = 0.0f;
        glEnable(GL_TEXTURE_2D);
        for (int i = 0; i < chars.length; i++) {
            float texPosX = chars[i] % 16 * glyphSize / (float)imageWidth;
            float texPosY = (chars[i] - 32) / 16 / 16.0f;
            float u1 = texPosX + glyphSize / (float)imageWidth;
            float v1 = texPosY + glyphSize / (float)imageWidth;
            glBegin(GL_QUADS);
            glTexCoord2f(texPosX, texPosY);
            glVertex3f(xPos + xOffset, yPos, 0.0f);
            glTexCoord2f(u1, texPosY);
            glVertex3f(xPos + xOffset + fontSize, yPos, 0.0f);
            glTexCoord2f(u1, v1);
            glVertex3f(xPos + xOffset + fontSize, yPos + fontSize, 0.0f);
            glTexCoord2f(texPosX, v1);
            glVertex3f(xPos + xOffset, yPos + fontSize, 0.0f);
            glEnd();
            xOffset += charWidths[chars[i]];
        }
    }

    public void drawBkg(int xPos, int yPos, int wdt, int color) {
        float r = (color >> 16 & 0xFF) / 255.0f;
        float g = (color >> 8  & 0xFF) / 255.0f;
        float b = (color       & 0xFF) / 255.0f;
        glColor4f(r, g, b, 0.2f);
        glDisable(GL_TEXTURE_2D);
        glEnable(GL_BLEND);
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
        glBegin(GL_QUADS);
        glVertex2f(xPos, yPos);
        glVertex2f(xPos + wdt, yPos);
        glVertex2f(xPos + wdt, yPos + 18.0f);
        glVertex2f(xPos, yPos + 18.0f);
        glEnd();
        glDisable(GL_BLEND);
    }

    public int width(String t) {
        char[] c = t.toCharArray();
        float n = 0.0f;
        for (int i = 0; i < c.length; i++) {
            n += charWidths[c[i]];
        }
        return (int)n;
    }

    public int getTextureId() {
        return textureId;
    }

    public void destroy(){
        if (textureId > 0){
            glDeleteTextures(textureId);
            textureId = 0;
        }
    }

}
